### DISCUSSION TITLE
work with opentelemetry without "http_route" and "http_url" attribute printed

### DISCUSSION THREAD

**Discussion by charzhao** [reactiveui/refit#1645] [General]
I use below code to get metrics with OTel:

.WithMetrics(cfg =>
            {
                cfg.AddAspNetCoreInstrumentation();
                cfg.AddHttpClientInstrumentation();
                cfg.AddRuntimeInstrumentation();
            });
but in the output i did not get http_route attribute as below:

http_client_request_duration_seconds_bucket{http_request_method="GET",http_response_status_code="200",network_protocol_version="1.1",server_address="****.com",url_scheme="https",le="0.5"} 0 1706514537848

Is there any way to make http_route and http_url here?

**Comment by NTDV** [comment]
Hello! Are there any solutions besides writing your own middleware?

**Comment by glennawatson** [comment]
This is how the OpenTelemetry HttpClient instrumentation behaves, not a Refit limitation. `AddHttpClientInstrumentation` does not emit `http.route` for client spans/metrics, because on the client side there is no server route template - it only knows the actual URL. `http.route` is a server-side semantic-convention attribute.

What you can do:

Refit attaches its `RestMethodInfo` (which carries the relative path template) to the outgoing `HttpRequestMessage` options under `HttpRequestMessageOptions.RestMethodInfo`. You can read that in the instrumentation's enrich callback and add your own attribute:

```csharp
cfg.AddHttpClientInstrumentation(o =>
{
    o.EnrichWithHttpRequestMessage = (activity, request) =>
    {
        if (request.Options.TryGetValue(
                new HttpRequestOptionsKey<RestMethodInfo>(HttpRequestMessageOptions.RestMethodInfo),
                out var info))
        {
            activity.SetTag("http.route", info.RelativePath);
        }
    };
});
```

(Adjust the property name to whatever `RestMethodInfo` exposes for the template in your version.)

For `http.url` / the full URL, that is controlled by the instrumentation's semantic-convention version and its URL redaction settings - check the OpenTelemetry.Instrumentation.Http options rather than Refit.

So: no built-in `http_route` for HTTP client metrics by spec; enrich the activity yourself using Refit's request options if you want it.
